[20.01] Run integration tests on github, fix various bugs (interactive tools, kubernetes, test framework)

mvdbeek opened the pull request:
We didn't run any of the tests that need docker on Jenkins (since the tests already run in docker).
This fixes:
  - Interactive tools on python 3
  - ssh/paramiko based tests and host key policy (they only worked on OSX)
  - Updates mulled_build API tests for new API
  - Fixes any_target_requires_extended_base logic
  - Many testing framework fixes

The only integration tests we still skip is test_kubernetes_staging.py ... for this we would need a means to spin up a local checkout of Galaxy in k8s. I'm hoping we can do this with galaxy-k8s in the future.

mvdbeek committed:
Run jenkins tests on github-ci

This allows running integration tests that require docker or kubernetes.
I've also broken up the tests to reduce runtime.

mvdbeek committed:
Bump paramiko version

mvdbeek committed:
Disable strict host key checking during tests

And fix parsing of shell_strict_host_key_check

mvdbeek committed:
Tweak paramiko missing hostkey policies

mvdbeek committed:
Adjust uid of testuser inside slurm/pbs container

mvdbeek committed:
Run ssh/cluster containers in privileged mode

mvdbeek committed:
Run fallback external_metadata when finishing job

That only affects the ShellJobRunner

mvdbeek committed:
Skip usermod on osx, isn't necessary with docker for mac

mvdbeek committed:
Fix interactivetool python 3 support and tests

- enables docker destination for interactivetool tests
- fixes a wrong argument in entrypoint management
- fixes the test setup that targeted a previous iteration (which was
  still called realtimetools)
- prevents deleting the working directory before python has a chance to
  start the container monitor (that's the sleep command)
- fixes some odd interactive tool test assertions
- uses unicode arguments in key_type_token_mapping.py.
  uwsgi arguments are always passed as bytestrings.

mvdbeek committed:
CliRunner lint fix

mvdbeek committed:
Update build_mulled integration test for new API

The previous API doesn't allow container installation after 2445425bf96a4f834cae961d18860f9a18a60c60

mvdbeek committed:
Fix mulled_build logic

mvdbeek committed:
Install pykube if kubectl on path

May perhaps want to limit this to when we know we run integration tests,
but I don't think this harms anything for now.

mvdbeek committed:
Run and fix grt/export.py test

It used to take 35 minutes on jenkins, but it doesn't anymore, as we're
using a clean database for each test class.

mvdbeek committed:
Fix kubernetes tests on minikube

mvdbeek committed:
Fix tool watcher unit tests; sleep needs to happen after watching the file

nsoranzo committed:
Use `@pytest.mark.skipif` decorator

mvdbeek committed:
Fix integration test case for database set through env var

mvdbeek committed:
Don't set GALAXY_CONFIG_OVERRIDE_DATABASE_CONNECTION if GALAXY_TEST_DB_URI is set

The actual database to target is generated dynamically, but since
GALAXY_CONFIG_OVERRIDE_DATABASE_CONNECTION takes precedence the
wrong database is used.

mvdbeek committed:
Use pytest.mark.skip, shows failure reason with -rxXs

mvdbeek committed:
Fix skip_unless_fixed_port

GALAXY_TEST_PORT is always set, GALAXY_TEST_PORT_RANDOM is set to 1
if GALAXY_TEST_PORT wasn't set externally.

mvdbeek committed:
Check conditional dependencies, even if no config file was found

mvdbeek committed:
Decorate class to skip, not class setup method

nsoranzo commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,48 @@
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        python-version: [3.7, 2.7]
```
Drop 2.7? It's 2020 :wink: 

```suggestion
        python-version: [3.7]
```

mvdbeek replied:
More than happy to, but I do want to see them green once at least.

nsoranzo commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,48 @@
+        path: .venv
+        key: cache-venv-galaxy-${{ matrix.python-version }}-${{ hashFiles('requirements.txt') }}
+    - name: Run tests
+      run: './run_tests.sh -integration test/integration -- -k "${{ matrix.subset }}"'
```
```suggestion
      run: './run_tests.sh -integration -- -k "${{ matrix.subset }}"'
```

mvdbeek replied:
That doesn't work:
```
./run_tests.sh -integration -- -k bla
invalid option: -k
```

nsoranzo commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,52 @@
+    strategy:
+      matrix:
+        python-version: [2.7]
+        subset: ['datatype']
```
```suggestion
        subset: ['upload_datatype']
```

mvdbeek replied:
Thanks, I've rebased this into the initial commit.

mvdbeek committed:
Print output when exception occurs

nsoranzo commented on scripts/grt/export.py:
```diff
@@ -146,7 +146,7 @@ def annotate(label, human_label=None):
 
     annotate('export_jobs_start', 'Exporting Jobs')
-    handle_job = io.open(REPORT_BASE + '.jobs.tsv', 'w', encoding='utf-8')
+    handle_job = codecs.open(REPORT_BASE + '.jobs.tsv', 'w', encoding='utf-8')
```
Not sure I understand the reason for this change, we are writing only unicodified content to this file handle.

mvdbeek replied:
On python 3, on python 2 with bytestring literals https://github.com/mvdbeek/galaxy/blob/05b253124ac32eca1521d11a40ad2bc6ff865755/scripts/grt/export.py#L150 and https://github.com/mvdbeek/galaxy/blob/05b253124ac32eca1521d11a40ad2bc6ff865755/scripts/grt/export.py#L184 and more instances of this cause problems

nsoranzo replied:
OMG, how this ever worked? Can I change this back to use `io` and change those lines to write Unicode?

mvdbeek replied:
Of course!

mvdbeek committed:
Just make sure we're not using galaxy's isolated home dir

nsoranzo committed:
Write unicode strings to files opened with `io.open(path, w, encoding=utf-8)`

Otherwise under Python2 the following exception is raised:
```
$ python2
Python 2.7.17 (default, Nov  7 2019, 10:07:09)
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> with io.open('foo.txt', 'w', encoding='utf-8') as fh:
...     fh.write('foo')
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: write() argument 1 must be unicode, not str
```

mvdbeek commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,48 @@
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        python-version: [3.7, 2.7]
```
```suggestion
        python-version: [3.7]
```
I guess something went wrong somewhere, I though I rebased it into the first commit ...

nsoranzo replied:
I temporarily dropped it to run the test on 2.7 too.

mvdbeek replied:
Oops, I was a bit too fast then

mvdbeek committed:
Drop 2.7 from integration tests

mvdbeek commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,48 @@
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        python-version: [3.7]
```
```suggestion
        python-version: [2.7, 3.7]
```

nsoranzo replied:
Test are still running, probably not needed.

nsoranzo commented:
Thanks a lot @mvdbeek for all the work on this!

Should we disable integration tests on Jenkins? For >=20.01 at least.

Also, do you plan to move other tests to GH workflows?

martenson commented:
@nsoranzo Unless we have other tests that we can't run on jenkins (or can run better on GHci) I think we are at a good spot without the need to move things around. The spread also makes our testing framework more robust I'd speculate.

mvdbeek commented:
Yes to disabling Jenkins integration tests on >= 20.01, otherwise I agree with @martenson.
Let's see how this service evolves, maybe it will make sense in the future to shift some nodes (especially the cloud workers from eu/au) to the [self-hosted runners](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/hosting-your-own-runners) to extend the capacity. The current capacity is limited to 20/40 concurrent jobs (20 for all, 40 for pro accounts), which I think we would exhaust too quickly if we moved all our testing jobs.

nsoranzo commented:
I've whitelisted the `release_19.05` and `release_19.09` target branches in the configuration of `docker-integration-py3` jobs on Jenkins: https://jenkins.galaxyproject.org/job/docker-integration-py3/jobConfigHistory/showDiffFiles?timestamp1=2019-11-19_05-54-42&timestamp2=2020-01-09_17-06-56

Let's see if this does the right thing!
